pkg/resource/restconfig_factory.go

```diff
@@ -2,30 +2,71 @@ package resource
 
 import (
 	controllerutils "github.com/openshift/hive/pkg/controller/utils"
+	log "github.com/sirupsen/logrus"
 	"k8s.io/apimachinery/pkg/api/meta"
 	"k8s.io/client-go/discovery"
 	"k8s.io/client-go/rest"
 	"k8s.io/client-go/restmapper"
 	"k8s.io/client-go/tools/clientcmd"
 	cmdutil "k8s.io/kubectl/pkg/cmd/util"
+	"sync"
 )
 
+// HIVE-2399: In the cluster sync controller, getRESTConfigFactory is called once per
+// reconcile loop. Previously, a restConfigClientGetter was created for every reconcile,
+// used to apply (or rather, attempt to apply) various synced objects, and then thrown
+// away at the end of the reconcile pass.
+//
+// The restConfigClientGetter (or more accurately, the  cmdutil.Factory that wraps it) is
+// used by the cluster sync controller (via this helper) to apply the various objects to
+// the target cluster during the syncing process. During the application process (see apply.go),
+// the apply command is set up (aptly named setupApplyCommand()) using the factory returned
+// by the getRESTConfigFactory() below. Within this setup, the ToRESTMapper() function of the
+// restConfigClientGetter is invoked. While the mapper returned is a deferred variant,
+// eventually the inner NewDiscoveryRESTMapper() function is invoked, which is very expensive
+// and allocates a huge amount of memory over time if repeatedly called in a hot loop. A graphic
+// demonstrating these allocations is attached the the HIVE-2399 card.
+//
+// The solution is twofold:
+// 1) store the deferred mapper on the restConfigClientGetter so that it can be reused
+// 2) cache the restConfigClientGetter for each rest config and reuse it for future reconciles
+var restConfigClientGetterCache sync.Map
```

Comment:
I'm not done reading yet, but unless we implement some kind of manual garbage collection, this cache will still be able to leak memory, as it will never remove an entry.

IIUC reconciling on a given spoke will require an entry, yes? But it could even be worse than that because if e.g. creds are rotated, that would result in a _different_ rest.Config for the same spoke?

And even if we managed to remove a "stale" entry, due to the nature of the underlying problem, it wouldn't mean we actually scrubbed the memory it consumed internally?

:arrow_up: partially-formed thoughts, discussion needed.